nixops/state/json_file.py

```diff
@@ -0,0 +1,356 @@
+                self._commit()
+            fcntl.flock(self._lock_file, fcntl.LOCK_UN)
+        self.lock.release()
+
+    def _rollback(self):
+        self._backup_state  = None
+        self._current_state = None
+        pass
+
+    def set(self, state):
+        self._current_state = state
+
+    def _commit(self):
+        assert self.nesting == 0
+
+        # TODO: write to temp file, then mv
+        with open(self.db_file, "w") as f:
+          json.dump(self._current_state, f,indent=2)
+
+        self._backup_state  = None
+        self._current_state = None
+
+    def _ensure_db_exists(self):
+        db_exists = os.path.exists(self.db_file)
+        if not db_exists:
+            initial_db = {
+              "schemaVersion": 0,
+              "deployments": {}
+            }
+
+            with open(self.db_file, "w", 0o600) as f:
+                json.dump(initial_db, f)
+                f.close()
+
+    def schema_version(self):
+        version = self.read()["schemaVersion"]
+        if version is None:
+            raise "illegal datafile" # TODO: proper exception
+        else:
+            return version
+
+class JsonFile(object):
+    """NixOps state file."""
+
+    def __init__(self, json_file):
+        self.file_path = json_file
+
+        if os.path.splitext(json_file)[1] not in ['.json']:
+            raise Exception("state file ‘{0}’ should have extension ‘.json’".format(json_file))
+
+        self.db = TransactionalJsonFile(json_file)
+
+        # Check that we're not using a new DB schema version.
+        with self.db:
+            version = self.db.schema_version()
+            if version > 0:
+               raise Exception("this NixOps version is too old to deal with JSON schema version {0}".format(version))
+
+    def close(self):
+        pass
+
+    ###############################################################################################
+    ## Deployment
+
+    def query_deployments(self):
+        """Return the UUIDs of all deployments in the database."""
+
+        return self.db.read()["deployments"].keys()
+
+    def get_all_deployments(self):
+        """Return Deployment objects for every deployment in the database."""
+        uuids = self.query_deployments()
+        res = []
+        for uuid in uuids:
+            try:
+                res.append(self.open_deployment(uuid=uuid))
+            except nixops.deployment.UnknownBackend as e:
+                sys.stderr.write("skipping deployment ‘{0}’: {1}\n".format(uuid, str(e)))
+        return res
+
+    def _find_deployment(self, uuid=None):
+        all_deployments = self.db.read()["deployments"]
+
+        found = []
+
+        # if nothing exists no reason to check for things
+        if not all_deployments:
+            return None
+
+        if not uuid:
+            found = filter(lambda(id): all_deployments[id]["attributes"].get("name"), all_deployments)
+        else:
+            found = filter(lambda(id): id == uuid, all_deployments)
```

Comment:
Please rewrite this logic in some less contrived way (please don't ask for details).

Reply:
@coretemp  done